paddle/fluid/framework/ir/multi_devices_graph_pass/backward_optimizer_op_deps_pass.cc

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "paddle/fluid/framework/details/container_cast.h"
+#include "paddle/fluid/framework/details/multi_devices_helper.h"
+#include "paddle/fluid/framework/details/op_handle_base.h"
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/graph_helper.h"
+#include "paddle/fluid/framework/ir/pass.h"
+#include "paddle/fluid/framework/lod_tensor.h"
+#include "paddle/fluid/framework/op_proto_maker.h"
+#include "paddle/fluid/framework/scope.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+class BackWardOpDepsPass : public ir::Pass {
+ protected:
+  void AddDep(ir::Graph* graph, details::OpHandleBase* l,
+              details::OpHandleBase* r) const {
+    auto* dep_var = new details::DummyVarHandle(graph->CreateControlDepVar());
+    graph->Get<details::GraphDepVars>(details::kGraphDepVars).emplace(dep_var);
+    l->AddOutput(dep_var);
+    r->AddInput(dep_var);
+    VLOG(10) << "add deps:" << l->DebugString() << " and " << r->DebugString();
+  }
+
+  void ApplyImpl(ir::Graph* graph) const override {
+    // NOTE: The operator nodes should be in topology order.
+    std::vector<details::OpHandleBase*> backward_op_handles;
+    std::vector<details::OpHandleBase*> all_opt_handles;
+    details::ParamsAndGrads params_grads;
+    std::vector<ir::Node*> topo_nodes = ir::TopologySortOperations(*graph);
+    for (auto& node : topo_nodes) {
+      if (!node->Op()) continue;
+
+      GetBackWardOpHandles(node, &backward_op_handles, &params_grads);
+      GetOptHandles(node, &all_opt_handles);
+    }
+
+    VLOG(10) << "backward_op_handles size:" << backward_op_handles.size()
+             << ", opt_handles size:" << all_opt_handles.size();
+
+    if (backward_op_handles.size() <= 1 || all_opt_handles.size() <= 1) {
+      VLOG(10) << "need not backward_op_deps_pass";
+      return;
+    }
+
+    std::vector<details::OpHandleBase*> opt_handles;
+    GetHeadOptHandles(all_opt_handles, &opt_handles, params_grads);
```

Comment:
What's the meaning of `GetHeadOptHandles `?
The name is confusing.

Reply:
Get the roots of the optimizer handles' graphs.

Reply:
Done.